### PR TITLE
Parse fields access

### PR DESCRIPTION
**Pull request by Override-6**
add field access parsing support :
```kt
{$self.b.method().field}.other_field
```

### CODE REVIEW

File: ast/src/variable.rs [hunk 1]
```diff
@@ -95,10 +112,10 @@ impl SourceSegmentHolder for Assign<'_> {
 }
 
 impl Assign<'_> {
-    pub fn name(&self) -> Option<&str> {
+    pub fn name(&self) -> Option<String> {
         match self.left.as_ref() {
-            Expr::Identifier(Identifier { name, .. }) => Some(name),
-            Expr::VarReference(VarReference { name, .. }) => Some(name),
+            Expr::Identifier(Identifier { name, .. }) => Some(name.to_string()),
+            Expr::VarReference(VarReference { name, .. }) => Some(name.to_string()),
```
Comment (syldium) [inline ast/src/variable.rs:118]:
```suggestion
            Expr::Identifier(Identifier { name, .. }) => Some(name),
            Expr::VarReference(VarReference { name, .. }) => Some(name.name()),
```

```rs
impl VarName<'_> {
    pub fn name(&self) -> &str {
        match self {
            VarName::User(n) => n,
            VarName::Slf => "self",
        }
    }
}
```

File: parser/src/aspects/call.rs [hunk 2]
```diff
@@ -113,20 +114,40 @@ impl<'a> CallAspect<'a> for Parser<'a> {
...
+            } else if self
+                .cursor
+                .lookahead(
+                    of_type(TokenType::Dot)
+                        .and_then(identifier_parenthesis())
+                        .or(of_types(&[
+                            TokenType::RoundedLeftBracket,
+                            TokenType::SquaredLeftBracket,
+                        ])),
+                )
```
Comment (syldium) [inline parser/src/aspects/call.rs:139]: This code is very similar to the one at the start of the method. Maybe start by parsing a field access, then see if there are parenthesis/square brackets.

File: ast/src/variable.rs [hunk 3]
```diff
@@ -35,12 +37,27 @@ pub enum VarKind {
     Val,
 }
 
+#[derive(Debug, Clone, PartialEq, DebugPls)]
+pub enum VarName<'a> {
+    User(&'a str),
+    Slf,
```
Comment (syldium) [inline ast/src/variable.rs:43]:
```suggestion
    /// An used defined variable name.
    User(&'a str),

    /// The `self` keyword to refer to the current object.
    Slf,
```

File: ast/src/lib.rs [hunk 4]
```diff
@@ -124,14 +125,15 @@ impl SourceSegmentHolder for Expr<'_> {
...
-            Expr::VarReference(var_reference) => var_reference.segment.clone(),
+            Expr::VarReference(var_reference) => var_reference.segment(),
             Expr::VarDeclaration(var_declaration) => var_declaration.segment.clone(),
             Expr::Range(range) => range.segment(),
             Expr::Subscript(subscript) => subscript.segment(),
             Expr::FunctionDeclaration(function_declaration) => function_declaration.segment.clone(),
             Expr::Parenthesis(parenthesis) => parenthesis.segment.clone(),
             Expr::Subshell(subshell) => subshell.segment.clone(),
             Expr::Block(block) => block.segment.clone(),
+            Expr::FieldAccess(fa) => fa.segment(),
```
Comment (syldium) [inline ast/src/lib.rs:136]: Group expression by their type.

**Review by syldium — CHANGES_REQUESTED** [review]

File: analyzer/src/steps/collect.rs [hunk 5]
```diff
@@ -391,14 +392,16 @@ impl<'a, 'b, 'e> SymbolCollector<'a, 'b, 'e> {
...
+                                    let symbol = self.identify_symbol(
+                                        *self.stack.last().unwrap(),
+                                        state.module,
+                                        SymbolLocation::unspecified(Name::new(name)),
+                                        reference.segment(),
+                                        SymbolRegistry::Objects,
+                                    );
+                                    self.current_env().annotate(reference, symbol);
+                                }
                             }
```
[the remark this answers is not in the document]
Comment (Override-6) [reply analyzer/src/steps/collect.rs:405]: this causes the match to have a default (`_ =>`) branch, not a huge fan of that

**Review by syldium — APPROVED** [review]

**Pull request merged**